tests/unit/s2n_tls_hybrid_prf_test.c

```diff
@@ -21,111 +21,86 @@
+
+    for (uint32_t i = 0; i < NUM_TEST_VECTORS; i++) {
+        /* Verify test index */
+        uint32_t count = 0;
+        GUARD(FindMarker(kat_file, "count = "));
+        gt_check(fscanf(kat_file, "%u", &count), 0);
+        eq_check(count, i);
+
         struct s2n_connection *conn;
         EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
         conn->actual_protocol_version = S2N_TLS12;
         /* Really only need for the hash function in the PRF */
         conn->secure.cipher_suite = &s2n_ecdhe_rsa_with_aes_256_gcm_sha384;
 
-        const char *premaster_classic_secret_hex_in = test_vectors[i][0];
-        const char *premaster_kem_secret_hex_in = test_vectors[i][1];
-        const char *client_random_hex_in = test_vectors[i][2];
-        const char *server_random_hex_in = test_vectors[i][3];
-        const char *client_key_exchange_message_1 = test_vectors[i][4];
-        const char *client_key_exchange_message_2 = test_vectors[i][5];
-        const char *expected_master_secret_hex_in = test_vectors[i][6];
+        /* Read test vector from KAT file */
+        uint32_t premaster_kem_secret_length = 0;
+        uint32_t client_key_exchange_message_length = 0;
+
+        GUARD(ReadHex(kat_file, premaster_classic_secret, PREMASTER_CLASSIC_SECRET_LENGTH, "premaster_classic_secret = "));
+
+        GUARD(FindMarker(kat_file, "premaster_kem_secret_length = "));
+        gt_check(fscanf(kat_file, "%u", &premaster_kem_secret_length), 0);
+
+        uint8_t *premaster_kem_secret;
+        notnull_check(premaster_kem_secret = malloc(premaster_kem_secret_length));
+        GUARD(ReadHex(kat_file, premaster_kem_secret, premaster_kem_secret_length, "premaster_kem_secret = "));
+
+        GUARD(ReadHex(kat_file, client_random, CLIENT_RANDOM_LENGTH, "client_random = "));
+        GUARD(ReadHex(kat_file, server_random, SERVER_RANDOM_LENGTH, "server_random = "));
+
+        GUARD(FindMarker(kat_file, "client_key_exchange_message_length = "));
+        gt_check(fscanf(kat_file, "%u", &client_key_exchange_message_length), 0);
+
+        uint8_t *client_key_exchange_message;
+        notnull_check(client_key_exchange_message = malloc(client_key_exchange_message_length));
+        GUARD(ReadHex(kat_file, client_key_exchange_message, client_key_exchange_message_length, "client_key_exchange_message = "));
+
+        GUARD(ReadHex(kat_file, expected_master_secret, MASTER_SECRET_LENGTH, "master_secret = "));
+
 
         DEFER_CLEANUP(struct s2n_blob classic_pms = {0}, s2n_free);
-        EXPECT_SUCCESS(s2n_alloc(&classic_pms, strlen(premaster_classic_secret_hex_in) / 2));
-        EXPECT_SUCCESS(convert_hex_to_bytes(classic_pms.data, premaster_classic_secret_hex_in));
+        EXPECT_SUCCESS(s2n_alloc(&classic_pms, PREMASTER_CLASSIC_SECRET_LENGTH));
+        for (uint32_t j = 0; j < PREMASTER_CLASSIC_SECRET_LENGTH; j++) {
+            classic_pms.data[j] = premaster_classic_secret[j];
+        }
```

Comment:
Since you already have the data in an array from `ReadHex` you could do: `struct s2n_blob classic_pms = {.data = premaster_classic_secret, .size = PREMASTER_CLASSIC_SECRET_LENGTH}` and get rid of this alloc/copy/free.

You could also do this for the `premaster_kem_secret` and randoms.

Reply:
Good call